components/psst/browser/content/psst_tab_helper.cc

```diff
@@ -0,0 +1,136 @@
+  DCHECK(psst_rule_service_);
+}
+
+PsstTabHelper::~PsstTabHelper() = default;
+
+void PsstTabHelper::OnTestScriptResult(
+    std::string policy_script,
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    base::Value value) {
+  if (value.GetIfBool().value_or(false)) {
+    InsertScriptInPage(render_frame_host_id, world_id, policy_script,
+                       base::DoNothing());
+  }
+}
+
+void PsstTabHelper::InsertTestScript(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    MatchedRule rule) {
+  CHECK(render_frame_host_id);
+  InsertScriptInPage(
+      render_frame_host_id, world_id, rule.test_script,
+      base::BindOnce(&PsstTabHelper::OnTestScriptResult,
+                     weak_factory_.GetWeakPtr(), rule.policy_script,
+                     render_frame_host_id, world_id));
+}
+
+void PsstTabHelper::InsertScriptInPage(
+    const content::GlobalRenderFrameHostId render_frame_host_id,
+    const int32_t world_id,
+    const std::string& script,
+    content::RenderFrameHost::JavaScriptResultCallback cb) {
+  content::RenderFrameHost* render_frame_host =
+      content::RenderFrameHost::FromID(render_frame_host_id);
+
+  // Check if render_frame_host is still valid and if starting rfh is the same.
+  if (render_frame_host &&
+      render_frame_host_id ==
+          web_contents()->GetPrimaryMainFrame()->GetGlobalId()) {
+    GetRemote(render_frame_host)
+        ->RequestAsyncExecuteScript(world_id, base::UTF8ToUTF16(script), false,
+                                    true, true, std::move(cb));
+  } else {
+    VLOG(2) << "render_frame_host is invalid.";
+    return;
+  }
+}
+
+mojo::AssociatedRemote<script_injector::mojom::ScriptInjector>&
+PsstTabHelper::GetRemote(content::RenderFrameHost* rfh) {
+  if (!script_injector_remote_.is_bound()) {
+    rfh->GetRemoteAssociatedInterfaces()->GetInterface(
+        &script_injector_remote_);
+  }
+  return script_injector_remote_;
+}
+
+void PsstTabHelper::DidFinishNavigation(
+    content::NavigationHandle* navigation_handle) {
+  if (!navigation_handle->IsInPrimaryMainFrame() ||
+      !navigation_handle->HasCommitted() ||
+      navigation_handle->IsSameDocument()) {
+    return;
+  }
+
+  DCHECK(psst_rule_service_);
+
+  should_process_ =
+      navigation_handle->GetRestoreType() == content::RestoreType::kNotRestored;
+}
+
+void PsstTabHelper::DocumentOnLoadCompletedInPrimaryMainFrame() {
+  DCHECK(psst_rule_service_);
+
+  if (!should_process_) {
+    return;
+  }
+  auto url = web_contents()->GetLastCommittedURL();
+
+  content::GlobalRenderFrameHostId render_frame_host_id =
+      web_contents()->GetPrimaryMainFrame()->GetGlobalId();
+
+  psst_rule_service_->CheckIfMatch(
+      url, base::BindOnce(&PsstTabHelper::InsertTestScript,
+                          weak_factory_.GetWeakPtr(), render_frame_host_id,
+                          world_id_));
```

Comment:
why do we need to pass `world_id_` here when it's a instance var that doesn't change? Can't we just pass it in when we call `RequestAsyncExecuteScript`?